csrc/iter_visitor.h

```diff
@@ -480,6 +488,21 @@ class DeadCodeRemover : BackwardVisitor {
   //! Mark a single Statement as being alive.
   inline void markLive(Statement* stmt) {
     live_statements_.insert(stmt);
+    if (auto e = dynamic_cast<Expr*>(stmt)) {
+      // Check if this expression is already in uses() for each of its inputs
+      // and if not, record it in future_uses_
+      for (Val* inp : e->inputs()) {
+        if (std::find(inp->uses().begin(), inp->uses().end(), e) ==
+            inp->uses().end()) {
+          auto fu_it = future_uses_.find(inp);
```

Comment:
how did we enter here?

so `e` has `inp` as its inputs but `inp` doesn't have `e` in its uses? we have a invalid graph here?

Reply:
The graph is valid, but `inp->uses()` only gives us uses **that lead to outputs**. In this function `stmt` will often be a replacement `Val` which is temporarily "dead" in that it hasn't yet been hooked up to a path leading to outputs.

Reply:
IDK whether we should leave this behavior of `uses()` or possibly change it since it's slightly surprising. It has been here a long time though: https://github.com/csarofeen/pytorch/commit/cfaecaf40bd6cabd3f4e0ef0d8c7252655349b61#diff-6e4fc28533e1790003083738b2994e3a44e45a05ff5264193d199357ff78baebR465-R488.

Reply:
Ha, thanks for bearing with my ignorance. `StmtSort::getStmts` adds an artificial end point `fusion->getTerminationOutputs()`.